cmd/ocsp-updater/main.go

```diff
@@ -107,97 +203,117 @@ func (updater *OCSPUpdater) processResponse(tx *gorp.Transaction, serial string)
 	return nil
 }
 
-// Produce one OCSP response for the given serial, returning err
-// if anything went wrong. This method will open and commit a transaction.
-func (updater *OCSPUpdater) updateOneSerial(serial string) error {
-	innerStart := time.Now()
-	// Each response gets a transaction. In the future we can increase
-	// performance by batching transactions.
-	// The key thing to think through is the cost of rollbacks, and whether
-	// we should rollback if CA/HSM fails to sign the response or only
-	// upon a partial DB insert.
-	tx, err := updater.dbMap.Begin()
-	if err != nil {
-		updater.log.Err(fmt.Sprintf("OCSP %s: Error starting transaction, aborting: %s", serial, err))
-		updater.stats.Inc("OCSP.Updates.Failed", 1, 1.0)
-		tx.Rollback()
-		// Failure to begin transaction is a fatal error.
-		return FatalError(err.Error())
+// newCertificateTick checks for certificates issued since the last tick and
+// generates and stores OCSP responses for these certs
+func (updater *OCSPUpdater) newCertificateTick(batchSize int, _, prev time.Time) {
+	updater.rpcMu.RLock()
+
+	// Check for anything issued between now and previous tick and generate the
+	// OCSP responses and submit them to CT logs using the Publisher
+	certs, err := updater.getCertificatesIssuedSince(prev, batchSize)
+	if err != nil && err != errNoNewCertificates {
+		updater.rpcMu.RUnlock()
+		return
 	}
 
-	if err := updater.processResponse(tx, serial); err != nil {
-		updater.log.Err(fmt.Sprintf("OCSP %s: Could not process OCSP Response, skipping: %s", serial, err))
-		updater.stats.Inc("OCSP.Updates.Failed", 1, 1.0)
-		tx.Rollback()
-		return err
-	}
+	// Do OCSP response generation / CT submission in parallel
+	wg := new(sync.WaitGroup)
+	updater.generateOCSPResponses(wg, certs)
+	wg.Wait()
+	updater.rpcMu.RUnlock()
+}
 
-	err = tx.Commit()
-	if err != nil {
-		updater.log.Err(fmt.Sprintf("OCSP %s: Error committing transaction, skipping: %s", serial, err))
+func (updater *OCSPUpdater) generateOCSPResponses(wg *sync.WaitGroup, certs []core.Certificate) {
+	responseChan := make(chan responseMeta)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for _, cert := range certs {
+			meta, err := updater.generateResponse(cert)
+			if err != nil {
+				updater.log.AuditErr(err)
+				continue
+			}
+			responseChan <- meta
+		}
+		close(responseChan)
+	}()
+	wg.Add(1)
+	go func() {
```

Comment:
this func doesn't need to be go'ed. You can do in this goroutine.

Also, per above, this work should not be done until all of the responses are generated.

You could do that with a single `chan []responseMeta` that was sent to instead of closed above, and pulled from here.